frontend/src/features/turnstile/useTurnstile.tsx

```diff
@@ -0,0 +1,152 @@
+import { useCallback, useEffect, useRef, useState } from 'react'
+import { useIntervalWhen } from 'rooks'
+
+import { featureFlags } from '~shared/constants'
+
+import { useScript } from '~hooks/useScript'
+
+import { useIsFeatureEnabled } from '~features/feature-flags/queries'
+
+type TurnstileBaseConfig = {
+  sitekey?: string
+  theme?: 'light' | 'dark' | 'auto'
+  size?: 'normal' | 'compact'
+  tabindex?: number
+  language?: string
+  execution?: 'default' | 'execute'
+  appearance?: 'always' | 'execute' | 'interaction-only'
+}
+
+interface UseTurnstileProps extends TurnstileBaseConfig {
+  // id of container to load Turnstile captcha in.
+  containerID?: string
+}
+
+type TurnstileConfig = TurnstileBaseConfig & {
+  callback?: (token: string) => void
+  'expired-callback'?: () => void
+  'error-callback'?: () => void
+}
+
+type TurnstileCallbacks = {
+  ready: (callback: () => void) => void
+  render: (containerID?: string, config?: TurnstileConfig) => string
+  reset: (widgetID: string | undefined) => void
+  execute: (widgetID: string) => void
+  getResponse: (widgetID: string) => string
+}
+
+declare global {
+  interface Window {
+    turnstile?: TurnstileCallbacks
+  }
+}
+
+// todo: check usages
+type TurnstileExecutionCallback = {
+  resolve?: (response: string | null) => void
+  reject?: (reason?: unknown) => void
+}
+
+const DEFAULT_CONTAINER_ID = 'turnstile-container'
+
+export const useTurnstile = ({
+  containerID = DEFAULT_CONTAINER_ID,
+  sitekey,
+  execution = 'execute',
+  theme = 'light',
+  appearance = 'interaction-only',
+}: UseTurnstileProps) => {
+  useScript(
+    'https://challenges.cloudflare.com/turnstile/v0/api.js?render=explicit',
```

Comment:
Could we double check what happens on the responder side if this endpoint is not reachable? 

**Context**
We had an issue previously where our users in China couldn't access `google.com` and the page failed to load completely. There was no workaround as the captcha is required to load even if it was not enabled by the admin.

Reply:
I'm afraid the page will fail to load if this endpoint is not reachable. This applies to Google reCAPTCHA too. AFAIK, captcha verification is a critical flow for forms since a submission is not possible if the captcha verification in the backend fails.

A possible approach I can think of is to guard both captchas behind a feature flag so that we can disable them completely if required but that might not be viable from a security POV.

Reply:
Oh, regarding your second point, if captcha is not enabled, we will not attempt to load this script